README.md

```diff
@@ -335,192 +357,227 @@ module with string literals may fail due to lack of memory resources,
+The maximum number of bytes that can be encoded at once by
+`string.encode` is 2<sup>31</sup>-1.  If an encoding would require more
+bytes, it is as if the codepoints can't be encoded (a trap).
+
+### Concatenation
 
 ```
-(string.advance_wtf8 str:stringref pos:i32 codepoints:i32)
-  -> pos:i32, codepoints:i32
-(string.rewind_wtf8 str:stringref pos:i32 codepoints:i32)
-  -> pos:i32, codepoints:i32
+(string.concat a:stringref b:stringref) -> stringref
 ```
-Return a new WTF-8 position for the string *`str`* which is
-*`codepoints`* codepoints after or before *`pos`*.  Also return the
-number of codepoints by which the new position differs from the old,
-which may be less than the *`codepoints`* parameter if end-of-string or
-beginning-of-string would be reached.  The *`codepoints`* argument is
-interpreted as a `u32`.
+Return a new stringref containing the codepoints from *`a`* followed by
+the codepoints from *`b`*.
 
-#### Invalid position values
+Note that the implementation should take care when, at any future time,
+treating the resulting string as a sequence of codepoints.  If *`a`*'s
+last codepoint is a high surrogate and *`b`*'s first codepoint is a low
+surrogate, these two codepoints combine into one, as if they were the
+two code units of a UTF-16-encoded unicode scalar value.
 
-All instructions which take string positions trap if the position is not
-valid for the given string.
+Concatenating two strings is a form of dynamic allocation and can fail.
+If an allocation fails, the implementation must trap.  Fallible
+`string.concat` is a possible future extension.
 
-#### Position value limits
+### Predicates
 
-The maximum valid `i32` position value is 2<sup>31</sup>.
+```
+(string.eq a:stringref b:stringref) -> i32
+```
+Return 1 if the strings *`a`* and *`b`* contain the same codepoint
+sequence.  Return 0 otherwise.
+
+```
+(string.is_usv_sequence str:stringref)
+  -> bool:i32
+```
+Return 1 if the string *`str`* is a sequence of unicode scalar values,
+and 0 otherwise.  A 0 result indicates that *`str`* contains isolated
+surrogates.
+
+## The `stringview` facility
 
-Evaluation of an instruction which would result in a WTF-16 position
-value greater than 2<sup>31</sup> must trap.  In practice, no web
-browser embedding allows for strings longer than 2<sup>31</sup>–1 WTF-16
-code units, so no string from JavaScript is too large for the
-instructions in this proposal, when processed as WTF-16.
+Three new reference types: `stringview_wtf8`, `stringview_wtf16`, and
+`stringview_iter`.  Opaque, like `externref` and `funcref`.
 
-Evaluation of an instruction which would result in a WTF-8 position
-value greater than 2<sup>31</sup> must trap.  Note that depending on the
-string contents, a position for a WTF-8 encoding may be numerically
-lesser or greater than the corresponding WTF-16 position.
+### `stringview_wtf8`
 
-Future extensions of the `stringref` proposal along the lines of the
-[memory64
+```
+(string.as_wtf8 str:stringref)
```

Comment:
Have you intentionally not included any form of position-based getter, which would return either the byte or the codepoint at that position? (That would make sense as `stringview_iter.cur` is the obvious alternative; just asking to clarify.)

More generally, it seems to me that the use cases for stringview_wtf8 are somewhat thin (prefix/suffix comparisons could be addressed with a `stringref.startswith` instruction; "streaming"/chunkwise encoding to linear memory doesn't seem all that useful when the original string has to reside in managed memory in its entirety), but if we assume that these use cases must be covered somehow, then it's indeed difficult to find alternatives for them, and I don't want to go in circles by suggesting to merge stringview_wtf8 into stringview_iter :-)
Maybe this particular question ("Is there a need for stringview_wtf8 or is stringview_iter enough?") should be the subject of an official poll of some larger audience? 
Put differently: you already state above (line 270) that stringviews are possibly post-MVP. Personally, I think the need for stringview_wtf16 is most obvious, and I'm also easily convinced that stringview.iter is good to have, so in my view stringview_wtf8 is the most likely candidate for post-MVP postponement.

Alternatively, we could make stringview_wtf8 more useful:
- drop string.measure_wtf8 in favor of stringview_wtf8.bytelength
- drop string.encode_{u,w}tf8 in favor of stringview_wtf8.encode (possibly a new variant that defaults pos=0 bytes=-1)

In particular: depending on engine implementation details there might not be a difference, but if an application wants to do a "measure, allocate, encode" sequence, then modeling this as "string.as_wtf8, stringview_wtf8.bytelength, ...linear memory allocation..., stringview_wtf8.encode" at least provides the possibility that the view creation can do the re-encoding, and the "encode" instruction can then be a `memcpy`; whereas string.measure_wtf8 is essentially forced to walk the whole string only for string.encode_wtf8 to have to individually inspect every codepoint once more.

Reply:
Yeah good questions here.

Just to clarify, the use case I have in mind for `stringview_wtf8` is a language that wants to represent strings as utf-8 / wtf-8, for example Rust.  Though I expect the initial implementation of stringref support in Rust to be an eager copy/encode to UTF-8 on the boundaries, I would expect that over time we'd see a desire to optimize specific interfaces to avoid a copy.  So for example you would might implement a prefix check as "compare the first 10 WTF-8 bytes to this UTF-8 sequence, which I already have in main memory because I'm Rust and that's how I like to do things".  The wtf8 stringview is intended to fill this use case.

Of course `stringview_wtf8` may be premature or not fit for purpose and I agree that it's a good idea to keep its presence in the MVP as an open question.